sale_loyalty_auto_refresh/models/sale_order.py

```diff
@@ -0,0 +1,153 @@
+        related="company_id.auto_refresh_coupon",
+    )
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        if self._check_skip_refresh():
+            return super().create(vals_list)
+
+        self_ctx = self.with_context(skip_auto_refresh_coupons=True)
+        orders = super(SaleOrder, self_ctx).create(vals_list)
+        orders._auto_refresh_coupons()
+        return orders
+
+    def write(self, vals):
+        if self._check_skip_refresh():
+            return super().write(vals)
+
+        old_data = self._read_recs_data()
+        self_ctx = self.with_context(skip_auto_refresh_coupons=True)
+        res = super(SaleOrder, self_ctx).write(vals)
+        new_data = self._read_recs_data()
+        # Until we restart Odoo, we won't get new triggers from params. Once restarted
+        # the method will return an empty set.
+        new_triggers = self._new_trigger()
+        if old_data != new_data or any(x in new_triggers for x in vals):
+            self._auto_refresh_coupons()
+        return res
+
+    def _auto_refresh_coupons(self):
+        orders = self.with_context(skip_auto_refresh_coupons=True).filtered(
+            type(self)._allow_recompute_coupon_lines
+        )
+        for order in orders:
+            order._update_programs_and_rewards()
+            order.action_apply_rewards()
+
+    def action_apply_rewards(self):
+        self.ensure_one()
+        claimable_rewards = self._get_claimable_rewards()
+        for coupon, reward in claimable_rewards.items():
+            try:
+                self._apply_program_reward(reward, coupon)
+                self._update_programs_and_rewards()
+            except Exception as e:
+                # Ignore exception errors to unblock the user when creating/writing
+                logger.debug(e)
+
+    def _allow_recompute_coupon_lines(self):
+        """Returns whether reward lines in order ``self`` can be recomputed
+        automatically.
+
+        Hook method, to be overridden for custom behaviours.
+
+        :return: True if the current SO allows automatic recomputation for
+        reward lines, False otherwise.
+        """
+        self.ensure_one()
+        return self.auto_refresh_coupon and self.state in ("draft", "sent")
+
+    @api.model
+    def _get_auto_refresh_coupons_triggers(self) -> set:
+        triggers = super()._get_auto_refresh_coupons_triggers()
+        triggers.update(
+            {
+                "order_line.auto_refresh_coupon_triggers_data",
+                "partner_id",
+            }
+        )
+        return triggers
+
+    def action_open_reward_wizard(self):
+        return super(
+            SaleOrder, self.with_context(skip_auto_refresh_coupons=True)
+        ).action_open_reward_wizard()
+
+    def _update_programs_and_rewards(self):
+        return super(
+            SaleOrder, self.with_context(skip_auto_refresh_coupons=True)
+        )._update_programs_and_rewards()
+
+
+class SaleOrderLine(models.Model):
```

Comment:
Could be worth it to move `SaleOrderLine` to its own file `sale_order_line.py`.

Reply:
Well I've seen different approaches regarding that. Some people prefer to keep it in the parent object file (`sale_order.py`). I am okay to keep like that. But move `SaleOrderLine` to his own file it's an option as well

Reply:
Yeah, usually if I have something like:
```
from odoo import fields, models


class MyModel(models.Model):
    _name = "my.model"

    name = fields.Char()
    line_ids = fields.One2many("my.model.line", "rec_id")


class MyModel(models.Model):
    _name = "my.model"

    descr = fields.Char()
    rec_id = fields.Many2one("my.model")
```
then it can make sense to have both classes in the same file.

But in this module's case, file's ~150 LoC already, and both classes contain business logic... might as well separate them for improved readability.